framework/src/main/java/org/checkerframework/framework/type/DefaultTypeHierarchy.java

```diff
@@ -267,97 +268,176 @@ protected boolean areEqualInHierarchy(
+            if (previousResult != null) {
+                return previousResult;
             }
-            return areEqualInHierarchy(inside, outside);
-        }
-
-        AnnotatedTypeMirror outsideUpperBound;
-        AnnotatedTypeMirror outsideLowerBound;
-        if (outside.getKind() == TypeKind.WILDCARD) {
-            outsideUpperBound = ((AnnotatedWildcardType) outside).getExtendsBound();
-            outsideLowerBound = ((AnnotatedWildcardType) outside).getSuperBound();
-        } else if (TypesUtils.isCaptured(outside.getUnderlyingType())) {
-            outsideUpperBound = ((AnnotatedTypeVariable) outside).getUpperBound();
-            outsideLowerBound = ((AnnotatedTypeVariable) outside).getLowerBound();
-        } else {
-            throw new BugInCF(
-                    "Expected a wildcard or captured type variable, but found " + outside);
+            typeargVisitHistory.add(inside, outside, currentTop, true);
+            boolean result = isContainedByBoundType(inside, new BoundType(outside), canBeCovariant);
+            typeargVisitHistory.add(inside, outside, currentTop, result);
+            return result;
         }
-        Boolean previousResult = typeargVisitHistory.result(inside, outside, currentTop);
-        if (previousResult != null) {
-            return previousResult;
+        if (canBeCovariant) {
+            return isSubtype(inside, outside, currentTop);
         }
-
-        typeargVisitHistory.add(inside, outside, currentTop, true);
-        boolean result =
-                isContainedWildcard(
-                        inside, outside, outsideUpperBound, outsideLowerBound, canBeCovariant);
-        typeargVisitHistory.add(inside, outside, currentTop, result);
-        return result;
+        return areEqualInHierarchy(inside, outside);
     }
 
-    private boolean isContainedWildcard(
-            AnnotatedTypeMirror inside,
-            AnnotatedTypeMirror outside,
-            AnnotatedTypeMirror outsideUpperBound,
-            AnnotatedTypeMirror outsideLowerBound,
-            boolean canBeCovariant) {
-
-        if (inside.equals(outside)) {
-            // If they are equal, outside always contains inside.
-            return true;
-        }
-
-        if (inside.getKind() == TypeKind.WILDCARD) {
-            outsideUpperBound =
-                    checker.getTypeFactory()
-                            .widenToUpperBound(outsideUpperBound, (AnnotatedWildcardType) inside);
-        }
-        while (outsideUpperBound.getKind() == TypeKind.WILDCARD) {
-            if (ignoreUninferredTypeArgument(outsideUpperBound)) {
-                return true;
+    /**
+     * Returns true if {@code outside} contains {@code inside}, that is, if the set of types denoted
+     * by {@code outside} is a superset of or equal to the set of types denoted by {@code inside}.
+     *
+     * <p>A declared type is considered a supertype of another declared type only if all of the type
```

Comment:
Rather than repeating this, I would refer to `isContainedBy()`.  You might be able to reduce the size of the previous paragraph too.